client/asset/interface.go

```diff
@@ -0,0 +1,135 @@
+	// penalized, the input still needs to be found so the swap can be completed.
+	// For BTC-like blockchains, every input of every block starting at the
+	// contract block will need to be scanned until the spending input is found.
+	// This could potentially be an expensive operation if performed long after
+	// the swap is broadcast. Realistically, though, the taker should start
+	// looking for the maker's redemption beginning at swapconf confirmations
+	// (at the latest) regardless of whether the server sends the 'redemption'
+	// message or not.
+	FindRedemption(ctx context.Context, txid string, vout uint32) (dex.Bytes, error)
+	// Refund refunds a contract. This can only be used after the time lock has
+	// expired.
+	Refund(Receipt) error
+	// Address returns an address for the exchange wallet.
+	Address() (string, error)
+	// Unlock unlocks the exchange wallet.
+	Unlock(pw string, dur time.Duration) error
+	// Lock locks the exchange wallet.
+	Lock() error
+}
+
+// Coin is some amount of spendable asset. Coin provides the information needed
+// to locate the unspent value on the blockchain.
+type Coin interface {
+	// ID is a unique identifier for this output, likely a txid.
+	ID() string
+	// Index is a secondary identifier needed to locate this coin.
+	Index() uint32
+	// Value is the available quantity, in atoms/satoshi.
+	Value() uint64
+	// Confirmations is the number of confirmations on this Coin's block. If the
+	// coin becomes spent, Confirmations should return an error.
+	Confirmations() (uint32, error)
+	// Redeem is any redeem script required to spend the coin.
+	Redeem() dex.Bytes
+}
+
+// Coins a collection of coins as returned by Fund.
+type Coins []Coin
+
+// Receipt holds information about a sent swap contract.
+type Receipt interface {
+	// Expiration is the time lock expiration.
+	Expiration() time.Time
+	// Coin is the contract output.
+	Coin() Coin
+}
+
+// AuditInfo is audit information about a swap contract needed to audit the
+// contract.
+type AuditInfo interface {
+	// Recipient is the string-encoded recipient address.
+	Recipient() string
+	// Expiration is the unix timestamp of the contract time lock expiration.
+	Expiration() time.Time
+	// Coin is the output that contains the contract.
+	Coin() Coin
+}
+
+// INPUT TYPES
+// The types below will be used by the client as inputs for the methods exposed
+// by the wallet.
+
+// SwapTx is the details needed to broadcast some number of swap contracts.
+type SwapTx struct {
+	// Inputs are the Coins being spent.
+	Inputs Coins
+	// Contracts are the SwapContracts to include in this SwapTx.
+	Contracts []*Contract
+}
+
+// Contract is a swap contract.
+type Contract struct {
+	// Address is the receiving address.
+	Address string
+	// Value is the amount being traded.
+	Value uint64
+	// SecretHash is the hash of the secret key.
+	SecretHash dex.Bytes
+	// LockTime is the contract lock time.
+	LockTime uint64
+}
```

Comment:
Perhaps the contract should include the contract tx and associated refund tx as well?

Reply:
This is an input structure, so the contract tx and refund tx don't exist yet when this is used. I do like the idea of creating the refund transaction when the swap transaction is created so that the wallet doesn't need to remain unlocked while waiting to issue a refund.

Reply:
This may be something to consider in the future. I'm not crazy about the idea, because it forces the app to become the owner of data that it doesn't know how to use or store. The app should be able to construct the inputs that the exchange wallet requires for a refund from it's in-memory or DB data, and either the wallet will remain unlocked or the user will have to enter their password to broadcast the refund. That said, I do see the benefit of having a pre-constructed redemption transaction, so I'm open to looking at this again sometime.

Reply:
This may be something to consider in the future. I'm all that crazy about the idea, because it forces the app to become the owner of data that it doesn't know how to use. The app should be able to construct the inputs that the exchange wallet requires for a refund from in-memory or DB data, and either the wallet will remain unlocked or the user will have to enter their password to broadcast the refund. That said, I do see the benefit of having a pre-constructed redemption transaction, so I'm open to looking at this again sometime.